theseus/optimizer/manifold_gaussian.py

```diff
@@ -0,0 +1,107 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from itertools import count
+from typing import Optional, Sequence
+
+import torch
+
+from theseus.geometry import Manifold
+
+
+class ManifoldGaussian:
+    _ids = count(0)
+
+    def __init__(
+        self,
+        mean: Sequence[Manifold],
+        precision: Optional[torch.Tensor] = None,
+        name: Optional[str] = None,
+    ):
+        self._id = next(ManifoldGaussian._ids)
+        if name:
+            self.name = name
+        else:
+            self.name = f"{self.__class__.__name__}__{self._id}"
+
+        dof = 0
+        for v in mean:
+            dof += v.dof()
+        self._dof = dof
+
+        self.mean = mean
+        if precision is None:
+            self.precision = torch.zeros(mean[0].shape[0], self.dof, self.dof).to(
+                dtype=mean[0].dtype, device=mean[0].device
+            )
+        else:
+            # internally checks dtype, device and shape
+            self.update(precision=precision)
+
+    @property
+    def dof(self) -> int:
+        return self._dof
+
+    @property
+    def device(self) -> torch.device:
+        return self.mean[0].device
+
+    @property
+    def dtype(self) -> torch.dtype:
+        return self.mean[0].dtype
+
+    # calls to() on the internal tensors
+    def to(self, *args, **kwargs):
+        for var in self.mean:
+            var = var.to(*args, **kwargs)
+        self.precision = self.precision.to(*args, **kwargs)
+
+    def copy(self, new_name: Optional[str] = None) -> "ManifoldGaussian":
+        if not new_name:
+            new_name = f"{self.name}_copy"
+        mean_copy = [var.copy() for var in self.mean]
+        return ManifoldGaussian(mean_copy, name=new_name)
+
+    def __deepcopy__(self, memo):
+        if id(self) in memo:
+            return memo[id(self)]
+        the_copy = self.copy()
+        memo[id(self)] = the_copy
+        return the_copy
+
+    def update(
+        self,
+        mean: Optional[Sequence[Manifold]] = None,
+        precision: Optional[torch.Tensor] = None,
```

Comment:
Why is precision and mean are both optional. Would there be cases where you just update mean but not precision or vice-versa?

Reply:
Probably there aren't any cases when you'll want to update one and not the other. I'll make both required arguments.